Split UserManagement into seperate Cluster and Namespace Roles and RoleBindings

Nokel81 opened the pull request:
- EditableList blocks event propogation when adding an item

- Clarifications on which values should be rendered between cluster and
  non-cluster RB and R

Fixes https://github.com/lensapp/lens/issues/2199

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Screenshots:
![Screen Shot 2021-05-31 at 8 57 54 AM](https://user-images.githubusercontent.com/8225332/120196895-6113d900-c1ee-11eb-8553-c8e3523e8dca.png)
![Screen Shot 2021-05-31 at 8 57 47 AM](https://user-images.githubusercontent.com/8225332/120196899-61ac6f80-c1ee-11eb-9128-9fb11852fae9.png)
![Screen Shot 2021-05-31 at 8 57 40 AM](https://user-images.githubusercontent.com/8225332/120196900-62450600-c1ee-11eb-81e1-4fd4309bcffd.png)


Needed work:

- [x] ~~Add back name editing.~~ (This shouldn't have been possible, `kubectl` blocks name editing).
- [x] Remove multi-api from `KubeApi` and subscribing.

Nokel81 committed:
Split User Management types into seperate places

- simplify subscribe
- fix NamespaceStore subscribe
- remove showClusterOption

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nachasic reviewed (dismissed):
That's a hefty amount of changes. Overall LGTM

nevalla reviewed (changes requested):
Got app crash when tried to create new RoleBinding:

![image](https://user-images.githubusercontent.com/455844/119639543-6228bd00-be20-11eb-9909-0aaef273041d.png)

Can't open ClusterRoleBinding details:
![image](https://user-images.githubusercontent.com/455844/119640246-14f91b00-be21-11eb-80b2-2bb78b619422.png)

When trying to create new ClusterRoleBinding nothing is happening when clicking + button.

Rules are not shown in ClusterRole details.

Also seeing some mobx errors
```
mobx] Encountered an uncaught exception that was thrown by a reaction or observer component, in: 'Reaction[Reaction@691]' Error: Unable to find node on an unmounted component.
    at findCurrentFiberUsingSlowPath (react-dom.development.js:4303)
    at findCurrentHostFiber (react-dom.development.js:4465)
    at findHostInstanceWithWarning (react-dom.development.js:25389)
    at findDOMNode (react-dom.development.js:26067)
    at Dialog.get elem [as elem] (dialog.tsx:74)
    at Dialog.value [as onClose] (dialog.tsx:128)
    at Dialog.close (dialog.tsx:109)
    at dialog.tsx:66
    at executeAction (mobx.esm.js:1128)
    at Reaction@691 (mobx.esm.js:1112)
    at reactionRunner (mobx.esm.js:2679)
    at Reaction.onInvalidate_ (mobx.esm.js:2660)
    at Reaction.runReaction_ (mobx.esm.js:2289)
    at runReactionsHelper (mobx.esm.js:2455)
    at reactionScheduler (mobx.esm.js:2430)
    at mobx.esm.js:2468
```

Nokel81 committed:
Fix Dialog not being openable, rearrange tabs

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Return this from static funcitons

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jakolehm commented on src/common/utils/hash-set.ts:
```diff
@@ -0,0 +1,147 @@
+  return iterator as IterableIterator<T>;
+}
+
+export class ObservableHashSet<T> implements Set<T>, IInterceptable<ISetWillChange>, IListenable {
```
Do we really need to invent these new types? What's wrong with the existing ones? If we really need these please write unit tests.

Nokel81 replied:
Will write some tests. I believe that this is a useful type since it acts like a `Set<T>` with custom hash function (where hashing is 1-to-1).

jakolehm commented on src/common/utils/hashers.ts:
```diff
@@ -0,0 +1,37 @@
+import type { RoleBindingSubject } from "../../renderer/api/endpoints";
+import type { KubeObject } from "../../renderer/api/kube-object";
+
+export function hashKubeObject(k: KubeObject): string {
```
This is used only in single place. Do we really need it as an util? (are we going to use this somewhere else)

Nokel81 replied:
This is used in two places, `src/renderer/components/+user-management/+cluster-roles-bindings/dialog.tsx` and `src/renderer/components/+user-management/+roles-bindings/dialog.tsx` but I can move it if you would like.

jakolehm commented on src/common/utils/hashers.ts:
```diff
@@ -0,0 +1,37 @@
+  return k.metadata.uid;
+}
+
+export function hashRoleBindingSubject(subject: RoleBindingSubject): string {
```
Same here, used only in one place... does it need to be an util?

jakolehm commented on src/common/utils/n-fircate.ts:
```diff
@@ -0,0 +1,47 @@
+ * @param parts What each array will be filtered to
+ * @returns A `parts.length` tuple of `T[]` where each array has matching `field` values
+ */
+export function nFircate<T>(from: Iterable<T>, field: keyof T, parts: [T[typeof field]]): [T[]];
```
Extremely hard to understand what's happening here. Needs unit tests if we really need this kind of util.

Nokel81 replied:
Will write some tests.

jakolehm commented on src/renderer/components/+namespaces/namespace.store.ts:
```diff
@@ -97,13 +97,16 @@ export class NamespaceStore extends KubeObjectStore<Namespace> {
 
-  getSubscribeApis() {
-    // if user has given static list of namespaces let's not start watches because watch adds stuff that's not wanted
+  subscribe() {
```
Why did we change this?

Nokel81 replied:
Removing `getSubscribeApis()` very much simplifies the `subscribe()` logic because it is not fixed at always 1 API.

jakolehm commented on src/renderer/components/+user-management/+roles-bindings/details.tsx:
```diff
@@ -19,35 +19,32 @@
-
-  constructor(props: Props) {
-    super(props);
-    makeObservable(this);
```
We don't need `makeObservable` because no `@observable`?

Nokel81 replied:
That is correct.

jakolehm commented on src/renderer/components/+user-management/+service-accounts/details.tsx:
```diff
@@ -19,27 +19,27 @@
 interface Props extends KubeObjectDetailsProps<ServiceAccount> {
 }
 
-@observer
```
Why observer removed?

jakolehm commented on src/renderer/components/app.scss:
```diff
@@ -29,7 +29,7 @@
   --padding: var(--unit);
   --margin: var(--unit);
-  --border-radius: 3px;
+  --border-radius: 5px;
```
This probably should be reverted, out-of-scope for this PR.

jakolehm commented on src/renderer/components/app.tsx:
```diff
@@ -20,57 +20,57 @@
  */
 import React from "react";
 import { observable, makeObservable } from "mobx";
+import { webFrame } from "electron";
```
Did you shuffle order of imports by accident? Quite hard to say what's going on with this file.

Nokel81 replied:
Yes, will revert.

jakolehm commented on src/renderer/components/editable-list/editable-list.scss:
```diff
@@ -23,7 +23,7 @@
     display: flex;
     flex-direction: column;
-    margin-top: $padding * 2;
+    margin: $padding 0px;
```
Any explanation why do we want to change EditableList styles?

Nokel81 replied:
The `*/details.tsx` files use `<EditableList>` so these changes we to make it compatible with several of them in a column.

jakolehm commented on src/renderer/components/editable-list/editable-list.tsx:
```diff
@@ -71,17 +65,15 @@ export class EditableList<T> extends React.Component<Props<T>> {
           {
             items.map((item, index) => (
-              <div key={`${item}${index}`} className="el-item Badge">
+              <div key={`${item}${index}`} className="el-item">
```
Why Badge?

Nokel81 replied:
Do you mean, why am I removing that `className`? Because I am now doing the css for these items directly.

jakolehm commented on src/renderer/components/editable-list/editable-list.tsx:
```diff
@@ -71,17 +65,15 @@ export class EditableList<T> extends React.Component<Props<T>> {
           <Input
-            theme="round-black"
-            value={this.currentNewItem}
+            theme="round"
```
Why different theme?

Nokel81 replied:
Because when used within a `<Wizard>` we don't want it to be dark, but I do want the round outlined version.

jakolehm commented on src/renderer/components/input/input.scss:
```diff
@@ -48,7 +48,7 @@
 
     position: relative;
-    padding: $padding /4 * 3 0;
+    padding: $padding / 4 * 3 0;
```
These are global changes to Input are out-of-scope for this PR. Please describe why these are needed (if they are needed).

Nokel81 replied:
They are needed to facilitate the usability of `<EditableList>` within `<Wizard>`. I have tried to make sure that there are no visual changes to the rest of the application.

jakolehm commented on src/renderer/components/input/input.tsx:
```diff
@@ -41,7 +41,7 @@ type InputElement = HTMLInputElement | HTMLTextAreaElement;
 
 export type InputProps<T = string> = Omit<InputElementProps, "onChange" | "onSubmit"> & {
-  theme?: "round-black";
+  theme?: "round-black" | "round";
```
Did we invent new theme, `round` here?

Nokel81 replied:
Yes, just like `round-black` isn't a theme that is selectable. (From the `ThemeStore`'s POV).

jakolehm commented on src/renderer/components/input/input.tsx:
```diff
@@ -90,7 +90,7 @@ export class Input extends React.Component<InputProps, State> {
   }
 
-  setValue(value: string) {
+  setValue(value = "") {
```
Why do we change logic of `Input`?

Nokel81 replied:
Because `<Wizard>` wraps its contents in a `<form>` and `<EditableList>` also uses the equivalent of `onEnter`, these changes are needed to make sure that the `<form>.onSubmit` isn't always called.

jakolehm commented on src/renderer/components/kube-object/kube-object-details.tsx:
```diff
@@ -99,11 +100,7 @@ export class KubeObjectDetails extends React.Component {
-    if (store) {
-      return store.getByPath(this.path);
-    }
+    return apiManager.getStore(this.path)?.getByPath(this.path);
```
Idk how changes within this file are related to PR. Looks better but are these added here by accident?

Nokel81 replied:
Yes by accident. I'll revert into a separate PR.

jakolehm commented on src/renderer/components/select/select.scss:
```diff
@@ -46,16 +46,15 @@ html {
 
     &__control {
-      border: none;
+      border: solid 1px $halfGray;
```
And again, we are modifying global component styles without any explanation. Probably out-of-scope for this PR?

Nokel81 replied:
I don't remember about this one. The other style changes I remember being needed for `<Wizard>` but I don't remember if this one way, will revert and check.

jakolehm commented on src/renderer/components/select/select.tsx:
```diff
@@ -65,8 +66,10 @@ export class Select extends React.Component<SelectProps> {
 
-  @computed get theme() {
-    return this.props.themeName || ThemeStore.getInstance().activeTheme.type;
+  @computed get themeClass() {
```
Why did we change `Select`? Out of scope for this PR, right?

Nokel81 replied:
There is one more bug with `<Select>` still (which I found while writing this PR). That when used within `<Wizard>` the background colour for "Selected and hover" is still the global theme.

jakolehm commented on src/renderer/kube-object.store.ts:
```diff
@@ -306,51 +306,31 @@ export abstract class KubeObjectStore<T extends KubeObject = any> extends ItemSt
-    return [this.api];
-  }
-
-  subscribe(apis = this.getSubscribeApis()) {
```
So we decided to change KubeObjectStore subscibe api in this PR? Why?

Nokel81 replied:
Because all stores now have only one API that they know about/need to know about. Since there are now separate stores for the `ClusterRoles` and `ClusterRoleBindings`.

If you really want I can move that to a separate PR, but I felt that it made sense to do it in the same PR that allowed for it to happen.

nevalla replied:
should we also modify signature of `ItemStore.subscribe()` as well? (separate PR is fine if needed)

Nokel81 replied:
So the signature of `ItemStore.subscribe()` is `subscribe(...args: any[])`. Which is rather too JS-like for my liking. I think other changes will have to be done before `ItemStore.subscribe()` can be changed, though TBH I think only `KubeObjectStore` overrides `subscribe()` so that might be a possible solution (for a follow up PR).

Nokel81 committed:
Revert global changes, rename paths to be correct

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jim-docker commented on src/common/utils/hash-set.ts:
```diff
@@ -0,0 +1,145 @@
+      other.forEach(value => this.add(value));
+    } else if (other instanceof Set || other instanceof ObservableHashSet || other instanceof ObservableSet) {
+      this.clear();
+      other.forEach(value => this.add(value));
```
Doing the same thing in both cases?

Nokel81 committed:
Add unit tests

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jim-docker commented on src/renderer/components/+user-management/+cluster-role-bindings/details.tsx:
```diff
@@ -0,0 +1,167 @@
+          <Table selectable className="bindings box grow">
+            <TableHead>
+              <TableCell checkbox />
+              <TableCell className="binding">Binding</TableCell>
```
The title of this drawer is Bindings so this cell header could be Name instead of Binding?

jim-docker commented on src/renderer/components/+user-management/+cluster-role-bindings/details.tsx:
```diff
@@ -0,0 +1,167 @@
+                  <TableRow
+                    key={i}
+                    selected={isSelected}
+                    onClick={prevDefault(() => this.selectSubject(subject))}
```
seems a bit inconsistent that clicking on the whole item selects it (instead of just the checkbox). Or is there a precedent for selection on a details page?

Nokel81 replied:
How that is normal (iirc).

jim-docker reviewed (commented):
tried to add a new binding from the role bindings details page, it added the binding but also duplicated the existing binding (user1). Tried to delete the binding but it stayed. Then couldn't open the details page (nor the vertical dots menu). This video shows some of the weirdness:

https://user-images.githubusercontent.com/40840436/120718027-9b9f9f00-c496-11eb-83c8-5aac374e73ae.mov

Nokel81 committed:
Several fixes:

- Not being able to type in the ClusterRole add dialog

- Editing not working as intended on subjects for CRBs and RBs

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
remove isClusterScoped

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Fix not being able to select ServiceAccounts

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jim-docker reviewed (commented):
Is this a bug?

https://user-images.githubusercontent.com/40840436/121082017-7ae38c00-c7ab-11eb-861b-af37b7768596.mov

Nokel81 committed:
fix CRB's not being able to specify names, fixed SA not being bound to their namespace

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Merge branch 'master' of github.com:lensapp/lens into split-user-management

nevalla commented on src/renderer/api/endpoints/role-binding.api.ts:
```diff
@@ -24,15 +24,17 @@ import { KubeObject } from "../kube-object";
+  kind: RoleBindingSubjectKind;
   name: string;
-  namespace?: string;
+  namespace: string;
```
does this mean that `Group` or `User` must always be namespaced?

Nokel81 replied:
For `RoleBinding` subjects yes, they need to be the same as the namespace that the `RoleBinding` is in.

nevalla replied:
> For `RoleBinding` subjects yes, they need to be the same as the namespace that the `RoleBinding` is in.

That's not exactly true, the namespace just indicates in what namespace the service account does exist. It does not need to be the same namespace where RoleBinding is created.

`user` and `group` does not need namespace, like this [example](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#rolebinding-example):
```
apiVersion: rbac.authorization.k8s.io/v1
# This role binding allows "jane" to read pods in the "default" namespace.
# You need to already have a Role named "pod-reader" in that namespace.
kind: RoleBinding
metadata:
  name: read-pods
  namespace: default
subjects:
# You can specify more than one "subject"
- kind: User
  name: jane # "name" is case sensitive
  apiGroup: rbac.authorization.k8s.io
roleRef:
  # "roleRef" specifies the binding to a Role / ClusterRole
  kind: Role #this must be Role or ClusterRole
  name: pod-reader # this must match the name of the Role or ClusterRole you wish to bind to
  apiGroup: rbac.authorization.k8s.io

Nokel81 replied:
Ah okay, will revert that. From my testing it seemed like it was always there.

jim-docker commented on src/renderer/components/+user-management/+cluster-role-bindings/dialog.tsx:
```diff
@@ -0,0 +1,283 @@
+          value={this.selectedRoleRef}
+          onChange={({ value }) => {
+            this.selectedRoleRef = value;
+            this.bindingName = this.selectedRoleRef.getName();
```
it shouldn't do this if `this.bindingName` has already been set (i.e.when the user has already entered a name in the "Binding Name" field). Hmm, I guess this would lead to problems if the user changes the cluster role. Maybe don't set the binding name to the role name till create is invoked (and it wasn't already set by the user)?

Nokel81 replied:
How about detecting if the name has been changed and only overriding it if it hasn't been?

jim-docker replied:
But if I pick roleRef foo and then pick roleRef bar the name would stay foo? 

Personally I think the name should be the first input field and be required to be specified by the user. No autofill based on the role.

Nokel81 replied:
So it is common for the name of the binding to be the same as the role/clusterrole.

No, if you pick roleRef `foo` and then pick roleRef `bar`. Then name would change to `bar` unless you had modified it.

nevalla reviewed (dismissed):
Tested and seems to work as expected! In general LGTM, added one question that we need to verify before merging.

jim-docker commented on src/renderer/components/+user-management/+role-bindings/dialog.tsx:
```diff
@@ -0,0 +1,305 @@
+              this.bindingNamespace = this.selectedRoleRef.getNs();
+            }
+            
+            this.bindingName = this.selectedRoleRef.getName();
```
same here

jim-docker commented on src/renderer/components/+user-management/+role-bindings/view.tsx:
```diff
@@ -0,0 +1,91 @@
+          renderTableContents={(binding: RoleBinding) => [
+            binding.getName(),
+            <KubeObjectStatusIcon key="icon" object={binding} />,
+            binding.getNs() || "-",
```
role bindings must be associated with a namespace so it'd never be `"-"`?

jim-docker commented on src/renderer/components/+user-management/+role-bindings/dialog.tsx:
```diff
@@ -0,0 +1,305 @@
+  render() {
+    const { ...dialogProps } = this.props;
+    const [action, nextLabel] = this.isEditing ? ["Edit", "Update"] : ["Add", "Create"];
+    const disableNext = !this.selectedRoleRef || !this.selectedBindings.length;
```
should it test for namespace selected too? (I got a server error when I tried to create a role binding without selecting a namespace)

Nokel81 replied:
Sounds good

jim-docker commented on src/renderer/components/+user-management/+roles/details.scss:
```diff
@@ -0,0 +1,21 @@
+.ClusterRoleDetails {
```
Should this be
```suggestion
.RoleDetails {
```
?

Nokel81 committed:
Fix review comments

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
minor textual changes

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Revert changing RB's namespace to required

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
revert

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Fix not being able to select roles and clusterroles

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jim-docker reviewed (approved):
This could be addressed in a followup PR, if necessary:

https://user-images.githubusercontent.com/40840436/121401303-8effc880-c926-11eb-9bbe-3566269b0f32.mov

For a binding created via the create dialog, then updated via the update dialog, this error message appears. Though the resource is probably fine, the dialog stays open, and pressing "Update" again (which is probably not necessary?) works without error.
